syntax = "proto3";
package logic;
import "gate.proto";

option cc_generic_services = true;


message PushMsg {
    enum Type {
        PUSH = 0;
        ROOM = 1;
        BROADCAST = 2;
    }
    Type type = 1;
    int32 operation = 2;
    int32 speed = 3;
    string server = 4;
    string room = 5;
    repeated string keys = 6;
    bytes msg = 7;
}

message CloseReply {
}

message CloseReq {
}

message PingReply {
}

message PingReq {
}

message ConnectReq {
    string server = 1;
    string cookie = 2;
    bytes token = 3;
}

message ConnectReply {
    int64 mid = 1;
    string key = 2;
    string roomID = 3;
    repeated int32 accepts = 4;
    int64 heartbeat = 5;
}

message DisconnectReq {
    int64 mid = 1;
    string key = 2;
    string server = 3;
}

message DisconnectReply {
    bool has = 1;
}

message HeartbeatReq {
    int64 mid = 1;
    string key = 2;
    string server = 3;
}

message HeartbeatReply {
}

message OnlineReq {
    string server = 1;
    map<string, int32> roomCount = 2;
}

message OnlineReply {
    map<string, int32> allRoomCount = 1;
}

message ReceiveReq {
    int64 mid = 1;
    gate.Proto proto = 2;
}

message ReceiveReply {
}

service Logic {
    // Ping Service 
    rpc Ping(PingReq) returns(PingReply);
    // Close Service 
    rpc Close(CloseReq) returns(CloseReply);
    // Connect
    rpc Connect(ConnectReq) returns (ConnectReply);
    // Disconnect
    rpc Disconnect(DisconnectReq) returns (DisconnectReply);
    // Heartbeat
    rpc Heartbeat(HeartbeatReq) returns (HeartbeatReply);
    // RenewOnline
    rpc RenewOnline(OnlineReq) returns (OnlineReply);
    // Receive
    rpc Receive(ReceiveReq) returns (ReceiveReply);
}